core/src/main/java/org/infinispan/persistence/marshallers/PersistenceMarshaller.java

```diff
@@ -0,0 +1,184 @@
+         ctx.registerMarshaller(new WrappedByteArrayMarshaller());
+         ctx.registerMarshaller(new ByteBufferMarshaller());
+         ctx.registerMarshaller(new NumericVersion.Marshaller());
+         ctx.registerMarshaller(new SimpleClusteredVersion.Marshaller());
+         ctx.registerMarshaller(new EntryVersionMarshaller());
+         ctx.registerMarshaller(new UserObject.Marshaller(userMarshaller));
+         ctx.registerMarshaller(new KeyValuePair.Marshaller(this));
+         ctx.registerMarshaller(new MapMarshaller(FastCopyHashMap.class));
+         ctx.registerMarshaller(new AtomicKeySetImpl.Marshaller(gcr));
+      } catch (IOException e) {
+         log.error("Exception thrown when reading 'persistence.proto'", e);
+         throw new CacheException(e);
+      }
+   }
+
+   @Override
+   public void stop() {
+   }
+
+   public SerializationContext getSerializationContext() {
+      return serializationContext;
+   }
+
+   @Override
+   public MediaType mediaType() {
+      return MediaType.APPLICATION_PROTOSTREAM;
+   }
+
+   @Override
+   protected ByteBuffer objectToBuffer(Object o, int estimatedSize) throws IOException {
+      try {
+         byte[] bytes = ProtobufUtil.toWrappedByteArray(getSerializationContext(), wrap(o));
+         return new ByteBufferImpl(bytes, 0, bytes.length);
+      } catch (java.io.NotSerializableException nse) {
+         // TODO do we still want this? I think it assumes too much about the configured user marshaller as it may not
+         // even support the Serializable interface. Replace with generic MarshallingException?
+         if (log.isDebugEnabled()) log.debug("Object is not serializable", nse);
+         throw new NotSerializableException(nse.getMessage(), nse.getCause());
+      }
+   }
+
+   @Override
+   public Object objectFromByteBuffer(byte[] buf, int offset, int length) throws IOException, ClassNotFoundException {
+      return unwrap(ProtobufUtil.fromWrappedByteArray(serializationContext, buf, offset, length));
+   }
+
+   @Override
+   public void writeObject(Object o, OutputStream out) throws IOException {
+      WrappedMessage.writeMessage(serializationContext, RawProtoStreamWriterImpl.newInstance(out), wrap(o));
+   }
+
+   @Override
+   public Object readObject(InputStream in) throws ClassNotFoundException, IOException {
+      return unwrap(WrappedMessage.readMessage(serializationContext, RawProtoStreamReaderImpl.newInstance(in)));
+   }
+
+   private Object wrap(Object o) {
+      return isInternalClass(o) ? o : new UserObject(o);
+   }
+
+   private Object unwrap(Object o) {
+      return o instanceof UserObject ? ((UserObject)o).object : o;
+   }
+
+   @Override
+   public boolean isMarshallable(Object o) {
+      return !isBlacklisted(o) && (isInternalClass(o) || isUserMarshallable(o));
+   }
+
+   private boolean isBlacklisted(Object o) {
+      Class clazz = o.getClass();
+      if (clazz.isArray())
```

Comment:
We check arrays, should we check Collections as well?

Reply:
The details are slightly hazy at this point as it was a while since I encountered the issue, but this check is required for how lambda's are sometimes passed through to the marshaller. I don't think it's necessary todo the same for collections as I don't know of a scenario where a collection of `SerializedLambda` etc will be passed.